src/mirrors.rs

```diff
@@ -0,0 +1,240 @@
+/// A mirrored PostgreSQL client.
+/// Packets arrive to us through a channel from the main client and we send them to the server.
+use std::cmp::{max, min};
+use std::time::Duration;
+
+use bytes::{Bytes, BytesMut};
+
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+use tokio::time::sleep;
+
+use crate::config::{Address, Role, User};
+use crate::errors::Error;
+use crate::messages::flush;
+use crate::pool::ClientServerMap;
+use crate::server::Server;
+use crate::stats::get_reporter;
+use log::{error, info, trace, warn};
+
+const MAX_CONNECT_RETRIES: u32 = 5;
+const MAX_SEND_RETRIES: u32 = 3;
+
+pub struct MirroredClient {
+    address: Address,
+    user: User,
+    database: String,
+    bytes_rx: Receiver<Bytes>,
+    disconnect_rx: Receiver<()>,
+    successful_sends_without_recv: u32,
+}
+
+impl MirroredClient {
+    async fn recv_if_neccessary(&mut self, server: &mut Server) -> Result<(), Error> {
+        /* We only receive a response from the server if we have successfully sent
+          5 messages on the current connection. We also send a flush message to gaurantee
```

Comment:
Curious why this is necessary?

Reply:
If we sent a sequence like `P`, `B`, `B`, `B`, `E`, receiving will not yield anything and `recv` will block forever. Sending `H` will force the server to send back messages like `1`, `2` and the actual data rows.

Reply:
This was replaced with an extra branch in the top-level `select!`